// ng-tagcanvas/esm2015/src/tag-canvas/tag-canvas.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ElementRef, ViewChild, Output, EventEmitter } from "@angular/core";
import { v4 as uuid } from 'uuid';
export class TagCanvasComponent {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this._instanceId = uuid();
        this.canvasId = `canvas-${this._instanceId}`;
        this.tagListId = `tags-${this._instanceId}`;
        this.defaultOptions = {
            textFont: null,
            textColour: null,
            weightFrom: 'data-weight' // TODO: Hide this field from external modules
        };
        this.tagClick = new EventEmitter();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set options(value) {
        this._options = value;
        this.start();
    }
    /**
     * @return {?}
     */
    get options() {
        return this._options;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        let /** @type {?} */ canvasElement = /** @type {?} */ (this.canvas.nativeElement);
        canvasElement.width = Math.round(canvasElement.clientWidth);
        canvasElement.height = Math.round(canvasElement.clientHeight);
        let /** @type {?} */ style = window.getComputedStyle(this.elementRef.nativeElement);
        this.defaultOptions.textHeight = parseFloat(style['font-size']);
        this.start();
    }
    /**
     * @return {?}
     */
    start() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            let /** @type {?} */ mergedOptions = Object.assign({}, this.defaultOptions, this._options);
            yield blinkEyes();
            TagCanvas.Start(this.canvasId, this.tagListId, mergedOptions);
        });
    }
    /**
     * @return {?}
     */
    update() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield blinkEyes();
            TagCanvas.Update(this.canvasId);
        });
    }
    /**
     * @return {?}
     */
    ngDoCheck() {
        let /** @type {?} */ tagListState = JSON.stringify(this.tags.map(({ text: text, weight: weight }) => [text, weight])); // TODO: Improve performance, use KeyValueDiffer/IterableDiffer
        if (this._previousTagListState !== tagListState) {
            this._previousTagListState = tagListState;
            this.update();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        TagCanvas.Delete(this.canvasId);
    }
}
TagCanvasComponent.decorators = [
    { type: Component, args: [{
                selector: 'tag-canvas',
                template: `<canvas #canvas id="{{ canvasId }}"></canvas>
<div id="{{ tagListId }}" class="tag-container">
	<a href="javascript:void(0)" *ngFor="let tag of tags"
		[attr.data-weight]="tag.weight" (click)="tagClick.next(tag)">{{ tag.text }}</a>
</div>
`,
                styles: [`:host{display:block}canvas{display:block;width:100%;height:100%}.tag-container{display:none}.tag-container>a{font-family:inherit!important;color:inherit!important;font-size:inherit!important;font-weight:inherit!important}`]
            },] },
];
/** @nocollapse */
TagCanvasComponent.ctorParameters = () => [
    { type: ElementRef }
];
TagCanvasComponent.propDecorators = {
    tags: [{ type: Input }],
    options: [{ type: Input }],
    tagClick: [{ type: Output }],
    canvas: [{ type: ViewChild, args: ['canvas',] }]
};
function TagCanvasComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    TagCanvasComponent.prototype._instanceId;
    /** @type {?} */
    TagCanvasComponent.prototype.canvasId;
    /** @type {?} */
    TagCanvasComponent.prototype.tagListId;
    /** @type {?} */
    TagCanvasComponent.prototype.defaultOptions;
    /** @type {?} */
    TagCanvasComponent.prototype.canvasWidth;
    /** @type {?} */
    TagCanvasComponent.prototype.canvasHeight;
    /** @type {?} */
    TagCanvasComponent.prototype._options;
    /** @type {?} */
    TagCanvasComponent.prototype.tags;
    /** @type {?} */
    TagCanvasComponent.prototype.tagClick;
    /** @type {?} */
    TagCanvasComponent.prototype.canvas;
    /** @type {?} */
    TagCanvasComponent.prototype._previousTagListState;
    /** @type {?} */
    TagCanvasComponent.prototype.elementRef;
}
/**
 * \@description Waits for change detection cycle
 * @return {?} Promise<void>
 */
function blinkEyes() {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        yield new Promise(resolve => setTimeout(() => resolve()));
    });
}
