// ng-tagcanvas/bundles/ng-tagcanvas.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('uuid'), require('jquery-tagcanvas/tagcanvas'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ng-tagcanvas', ['exports', '@angular/core', 'uuid', 'jquery-tagcanvas/tagcanvas', '@angular/common'], factory) :
    (factory((global['ng-tagcanvas'] = {}),global.ng.core,null,null,global.ng.common));
}(this, (function (exports,core,uuid,tagcanvas,common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var TagCanvasComponent = (function () {
        function TagCanvasComponent(elementRef) {
            this.elementRef = elementRef;
            this._instanceId = uuid.v4();
            this.canvasId = "canvas-" + this._instanceId;
            this.tagListId = "tags-" + this._instanceId;
            this.defaultOptions = {
                textFont: null,
                textColour: null,
                weightFrom: 'data-weight' // TODO: Hide this field from external modules
            };
            this.tagClick = new core.EventEmitter();
        }
        Object.defineProperty(TagCanvasComponent.prototype, "options", {
            get: /**
             * @return {?}
             */ function () {
                return this._options;
            },
            // TODO: Add stretch (boolean) input
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) {
                this._options = value;
                this.start();
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        TagCanvasComponent.prototype.ngAfterViewInit = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ canvasElement = (this.canvas.nativeElement);
                canvasElement.width = Math.round(canvasElement.clientWidth);
                canvasElement.height = Math.round(canvasElement.clientHeight);
                var /** @type {?} */ style = window.getComputedStyle(this.elementRef.nativeElement);
                this.defaultOptions.textHeight = parseFloat(style['font-size']);
                this.start();
            };
        /**
         * @return {?}
         */
        TagCanvasComponent.prototype.start = /**
         * @return {?}
         */
            function () {
                return __awaiter(this, void 0, void 0, function () {
                    var mergedOptions;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                mergedOptions = Object.assign({}, this.defaultOptions, this._options);
                                return [4 /*yield*/, blinkEyes()];
                            case 1:
                                _a.sent();
                                TagCanvas.Start(this.canvasId, this.tagListId, mergedOptions);
                                return [2 /*return*/];
                        }
                    });
                });
            };
        /**
         * @return {?}
         */
        TagCanvasComponent.prototype.update = /**
         * @return {?}
         */
            function () {
                return __awaiter(this, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0: return [4 /*yield*/, blinkEyes()];
                            case 1:
                                _a.sent();
                                TagCanvas.Update(this.canvasId);
                                return [2 /*return*/];
                        }
                    });
                });
            };
        /**
         * @return {?}
         */
        TagCanvasComponent.prototype.ngDoCheck = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ tagListState = JSON.stringify(this.tags.map(function (_a) {
                    var text = _a.text, weight = _a.weight;
                    return [text, weight];
                })); // TODO: Improve performance, use KeyValueDiffer/IterableDiffer
                if (this._previousTagListState !== tagListState) {
                    this._previousTagListState = tagListState;
                    this.update();
                }
            };
        /**
         * @return {?}
         */
        TagCanvasComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                TagCanvas.Delete(this.canvasId);
            };
        TagCanvasComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'tag-canvas',
                        template: "<canvas #canvas id=\"{{ canvasId }}\"></canvas>\n<div id=\"{{ tagListId }}\" class=\"tag-container\">\n\t<a href=\"javascript:void(0)\" *ngFor=\"let tag of tags\"\n\t\t[attr.data-weight]=\"tag.weight\" (click)=\"tagClick.next(tag)\">{{ tag.text }}</a>\n</div>\n",
                        styles: [":host{display:block}canvas{display:block;width:100%;height:100%}.tag-container{display:none}.tag-container>a{font-family:inherit!important;color:inherit!important;font-size:inherit!important;font-weight:inherit!important}"]
                    },] },
        ];
        /** @nocollapse */
        TagCanvasComponent.ctorParameters = function () {
            return [
                { type: core.ElementRef }
            ];
        };
        TagCanvasComponent.propDecorators = {
            tags: [{ type: core.Input }],
            options: [{ type: core.Input }],
            tagClick: [{ type: core.Output }],
            canvas: [{ type: core.ViewChild, args: ['canvas',] }]
        };
        return TagCanvasComponent;
    }());
    /**
     * \@description Waits for change detection cycle
     * @return {?} Promise<void>
     */
    function blinkEyes() {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, new Promise(function (resolve) { return setTimeout(function () { return resolve(); }); })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var TagCanvasModule = (function () {
        function TagCanvasModule() {
        }
        TagCanvasModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [
                            TagCanvasComponent
                        ],
                        imports: [
                            common.CommonModule
                        ],
                        exports: [
                            TagCanvasComponent
                        ]
                    },] },
        ];
        return TagCanvasModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var Tag = (function () {
        function Tag(init) {
            Object.assign(this, init);
        }
        return Tag;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.TagCanvasModule = TagCanvasModule;
    exports.TagCanvasComponent = TagCanvasComponent;
    exports.Tag = Tag;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
